### DISCUSSION TITLE
Load jobs dynamically through configuration - as a plugin

### DISCUSSION THREAD

**Discussion by michaellwest** [quartznet/quartznet#889] [General] [answered]
Using the latest 3.1.0-beta2 is there a way to load jobs and triggers dynamically from referenced libraries? 

Looking for a way to allow other teams to build their custom jobs as a standalone library and drop into the application root.

**Comment by lahma** [comment, marked as answer]
Maybe it's the XML plugin that you are after? 

* [Here's an example how to do it just with regular configuration](https://github.com/quartznet/quartznet/blob/master/src/Quartz.Examples/15_ConfigureJobSchedulingByUsingXmlConfigurations/ConfigureJobSchedulingByUsingXmlConfigurationsExample.cs)
* [And here's an example with the new IoC configuration + typed helpers](https://github.com/quartznet/quartznet/blob/master/src/Quartz.Examples.AspNetCore/Startup.cs#L102-L109)
* [And the actual sample XML file](https://github.com/quartznet/quartznet/blob/master/src/Quartz.Examples.AspNetCore/quartz_jobs.config)

XML file can be set the be polled. You naturally need to "hard code" the types there, but as long as DLLs are in the same directory as the process, it should work. You can also reference the schema in repo to get intellisense.

**Comment by michaellwest** [comment]
Thank you @lahma I believe that is what I was looking for. I referenced the Quartz.Plugins project to add that.

Does this work with cron schedules too?

**Comment by lahma** [comment]
Yes, it supports cron indeed. Here's the [sample data folder used in tests](https://github.com/quartznet/quartznet/tree/master/src/Quartz.Tests.Integration/Xml/TestData). [Here's relevant lines for cron configuration](https://github.com/quartznet/quartznet/blob/master/src/Quartz.Tests.Integration/Xml/TestData/RichConfiguration_20.xml#L64-L89), the XSD should help you out to get intellisense available in your editor.

**Comment by lahma** [comment]
Closing as resolved.

**Comment by michaellwest** [comment]
Looks like durable and recover are invalid options for cron. 

https://github.com/quartznet/quartznet/blob/85da16a3d2a9a348c9ba7bb63f71b3f82573d310/src/Quartz.Tests.Integration/Xml/TestData/ComplexCron.xml#L11-L12

![image](https://user-images.githubusercontent.com/933163/87858261-485e4580-c8f2-11ea-8b27-ab01ba5b2582.png)

Is it also expected that when you specify a start and end time for cron that it only runs once?

When I remove this it runs as expected (`<cron-expression>0/10 * * ? * *</cron-expression>`)

```xml
<start-time>1982-06-28T18:15:00.0Z</start-time>
<end-time>2040-05-04T18:13:51.0Z</end-time>
```

**Comment by lahma** [comment]
Unfortunately the ordering matters (based on Java version's schema). Start and end time should work as expected, but if you don't need them, you shouldn't specify them.

**Comment by michaellwest** [comment]
Thanks again for your support. I built an application which loads unreferenced DLLs and I managed to get it to work with dependency injection using Scrutor. The XML plugin works nicely and kept me from having to build my own. I may still do that just so that I can learn how to build a plugin and load JSON files.

**Comment by lahma** [comment]
Great to hear, please consider contributing JSON support if you can make work like the XML version, a small refactor might be needed to feed a model that works for both.
